x-pack/auditbeat/processors/add_session_metadata/add_session_metadata.go

```diff
@@ -0,0 +1,192 @@
+		return ev, nil //nolint:nilerr // Running on events without PID is expected
+	}
+
+	err = p.provider.UpdateDB(ev)
+	if err != nil {
+		return ev, err
+	}
+
+	result, err := p.enrich(ev)
+	if err != nil {
+		return ev, fmt.Errorf("enriching event: %w", err)
+	}
+	return result, nil
+}
+
+func (p *addSessionMetadata) String() string {
+	return fmt.Sprintf("%v=[backend=%s, pid_field=%s, override_fields=%t]",
+		processorName, p.config.Backend, p.config.PidField, p.config.ReplaceFields)
+}
+
+func (p *addSessionMetadata) enrich(ev *beat.Event) (*beat.Event, error) {
+	pidIf, err := ev.GetValue(p.config.PidField)
+	if err != nil {
+		return nil, err
+	}
+	pid, err := pidToUInt32(pidIf)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse pid field '%s': %w", p.config.PidField, err)
+	}
+
+	fullProcess, err := p.db.GetProcess(pid)
+	if err != nil {
+		return nil, fmt.Errorf("pid %v not found in db: %w", pid, err)
+	}
+
+	result := ev.Clone()
+
+	processMap := fullProcess.ToMap()
+
+	err = mapstr.MergeFieldsDeep(result.Fields["process"].(mapstr.M), processMap, true)
+	if err != nil {
+		return nil, fmt.Errorf("merging enriched fields with event: %w", err)
+	}
+
+	if p.config.ReplaceFields {
+		if err := p.replaceFields(result); err != nil {
+			return nil, fmt.Errorf("replace fields: %w", err)
+		}
+	}
+	return result, nil
+}
+
+// pidToUInt32 converts PID value to uint32
+func pidToUInt32(value interface{}) (pid uint32, err error) {
+	switch v := value.(type) {
+	case string:
+		nr, err := strconv.Atoi(v)
+		if err != nil {
+			return 0, fmt.Errorf("error converting string to integer: %w", err)
+		}
+		pid = uint32(nr)
+	case uint32:
+		pid = v
+	case int, int8, int16, int32, int64:
+		pid64 := reflect.ValueOf(v).Int()
+		if pid = uint32(pid64); int64(pid) != pid64 {
+			return 0, fmt.Errorf("integer out of range: %d", pid64)
+		}
+	case uint, uintptr, uint8, uint16, uint64:
+		pidu64 := reflect.ValueOf(v).Uint()
+		if pid = uint32(pidu64); uint64(pid) != pidu64 {
+			return 0, fmt.Errorf("integer out of range: %d", pidu64)
+		}
+	default:
+		return 0, fmt.Errorf("not an integer or string, but %T", v)
+	}
+	return pid, nil
+}
+
+// replaceFields replaces event fields with values suitable user with the session viewer in Kibana
+// The current version of session view in Kibana expects different values than what are used by auditbeat
+// for some fields. This function converts these field to have values that will work with session view.
+//
+// This function is temporary, and can be removed when Kibana is updated to work with the auditbeat field values.
+func (p *addSessionMetadata) replaceFields(ev *beat.Event) error {
+	kind, err := ev.Fields.GetValue("event.kind")
+	if err != nil {
+		return err
+	}
+	isAuditdEvent, err := ev.Fields.HasKey("auditd")
+	if err != nil {
+		return err
+	}
+	if kind == "event" && isAuditdEvent {
+		// process start
+		syscall, err := ev.Fields.GetValue("auditd.data.syscall")
+		if err != nil {
+			return err
```

Comment:
Should it be an error if an auditd event is not associated to a syscall? I'm not sure that all events contain a syscall (like a `type=ADD_USER`).

Reply:
Yes, I've changed this to return nil to account for other event types